### PR TITLE
added telemetry mechanism

### PR DESCRIPTION
**Pull request by radu1122**
# Pull Request Template

## Type of change

- [x] 🍕 New feature
- [ ] 🐛 Bug Fix
- [ ] 🔥 Breaking change
- [ ] 🧑‍💻 Improvement
- [ ] 📝 Documentation Update

## Description
Added telemetry system with anonymous data sending for genezio CLI. Also a env variable `GENEZIO_NO_TELEMETRY` to disable this feature.

## Checklist

- [x] My code follows the contributor guidelines of this project;
- [x] I have updated the documentation;
- [x] I have added tests;
- [x] New and existing unit tests pass locally with my changes;

### CODE REVIEW

File: src/telemetry/sdk/remote.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    if (typeof process !== "undefined" && process.versions != null && process.versions.node != null) {
+        const httpModule = 'http';
+        http = await import(httpModule);
+        const httpsModule = 'https';
+        https = await import(httpsModule);
+    }
+    importDone = true;
+}
+
+async function makeRequestBrowser(request: any, url: any) {
```
Comment (costinsin) [inline src/telemetry/sdk/remote.ts:20]:
```suggestion
async function makeRequestBrowser(request: any, url: string) {
```
Shouldn't this be a string?

File: src/telemetry/sdk/remote.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        req.end();
+    });
+}
+
+/**
+ * The class through which all request to the Genezio backend will be passed.
+ * 
+ */
+ export class Remote {
+    url: any = undefined;
```
Comment (costinsin) [inline src/telemetry/sdk/remote.ts:78]:
```suggestion
    url?: string= undefined;
```

File: src/telemetry/sdk/remote.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    agent: any = undefined;
```
Comment (costinsin) [inline src/telemetry/sdk/remote.ts:79]:
```suggestion
    agent?: Agent = undefined;
```

File: src/telemetry/session.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { readUTF8File, writeToFile } from "./../utils/file.js";
+import os from "os";
+import path from "path";
+import fs from "fs";
+import { debugLogger } from '../utils/logging.js';
+
+export async function getTelemetrySessionId(): Promise<string|undefined> {
+    const homeDirectory = os.homedir();
+    const loginConfigFilePath = path.join(homeDirectory, ".geneziotelemetryrc")
```
Comment (costinsin) [inline src/telemetry/session.ts:9]: Are we planning to only store the session ID in this file?

File: src/telemetry/telemetry.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      const newSessionId = uuidv4();
+      debugLogger.debug(`[GenezioTelemetry]`, `New session id: ${newSessionId}`);
+      saveTelemetrySessionId(newSessionId);
+      return newSessionId;
+    }
+    
+    return sessionId;
+  }
+
+  public static async sendEvent(eventType: string, cl?: string, errorTrace?: string): Promise<void> {
```
Comment (costinsin) [inline src/telemetry/telemetry.ts:20]:
```suggestion
  public static async sendEvent(eventType: string, cloudProvider?: string, errorTrace?: string): Promise<void> {
```

I find using short name versions hard to read. Is this intentional? If not, I think we should use more suggestible naming. This comment also applies to `ul`, `opS` and other more to come.

**Review by costinsin — COMMENTED** [review]
Looks good for me. Nice work. I have some little suggestions and questions about the code.

File: src/telemetry/sdk/remote.ts (hunk 1 above)
Comment (radu1122) [reply src/telemetry/sdk/remote.ts:20]: this is coming from genezio sdk

File: src/telemetry/sdk/remote.ts (hunk 2 above)
Comment (radu1122) [reply src/telemetry/sdk/remote.ts:78]: same here

File: src/telemetry/sdk/remote.ts (hunk 3 above)
Comment (radu1122) [reply src/telemetry/sdk/remote.ts:79]: same here

File: src/telemetry/session.ts (hunk 4 above)
Comment (radu1122) [reply src/telemetry/session.ts:9]: yes

File: src/telemetry/telemetry.ts (hunk 5 above)
Comment (radu1122) [reply src/telemetry/telemetry.ts:20]: i think you are right. i will change to full names

File: src/telemetry/sdk/remote.ts (hunk 1 above)
Comment (vladiulianbogdan) [reply src/telemetry/sdk/remote.ts:20]: We should remove these anys and put actual types. Let's make a task with all the improvements suggested by Costin.

File: src/telemetry/session.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { readUTF8File, writeToFile } from "./../utils/file.js";
+import os from "os";
+import path from "path";
+import fs from "fs";
+import { debugLogger } from '../utils/logging.js';
+
+export async function getTelemetrySessionId(): Promise<string|undefined> {
```
Comment (vladiulianbogdan) [inline src/telemetry/session.ts:7]: I think session id is a wrong name for this concept.

File: src/telemetry/telemetry.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      debugLogger.debug(`[GenezioTelemetry]`, `New session id: ${newSessionId}`);
+      saveTelemetrySessionId(newSessionId);
+      return newSessionId;
+    }
+    
+    return sessionId;
+  }
+
+  public static async sendEvent(eventType: string, cloudProvider?: string, errorTrace?: string): Promise<void> {
+    if (process.env.GENEZIO_NO_TELEMETRY == "1") {
```
Comment (vladiulianbogdan) [inline src/telemetry/telemetry.ts:21]: We should think where we document this to let the users know that they can disabled.

File: src/telemetry/telemetry.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      const newSessionId = uuidv4();
+      debugLogger.debug(`[GenezioTelemetry]`, `New session id: ${newSessionId}`);
+      saveTelemetrySessionId(newSessionId);
+      return newSessionId;
+    }
+    
+    return sessionId;
+  }
+
+  public static async sendEvent(eventType: string, cloudProvider?: string, errorTrace?: string): Promise<void> {
```
Comment (vladiulianbogdan) [inline src/telemetry/telemetry.ts:20]: eventType should be an enum.

**Pull request merged**